submissions/factorio_draftsman/main.md

```diff
@@ -0,0 +1,211 @@
+
+In an effort to keep this article brief, I'm not going to go too far into the nitty-gritty about the module or how exactly it works; I've spent a lot of time writing documentation for that purpose instead. Rather, I'm going to show off a couple of things that I've made with it so far, as well as potential things that could be made in the future, to try to illustrate exactly why I spent all this time making it in the first place.
+
+### Automatic Item Stack Sizes
+
+Oftentimes, for you want to figure out how much storage you need for a specific amount of an item. However, storage in Factorio is based on slots, not amounts, so the amount of storage is actually dependent on not only the amount you're trying to store, but also the stack size of that item. [You can design a circuit contraption](https://factorioprints.com/view/-LwH68dlvOEIKWkU5JqR) to determine this amount of slots that you need for any input item, but you'd need a big cell of combinators documenting the item sizes. This is *boring* to make, and easily broken if a new item is added or the stack size changes. Clearly, for such a simple and repetitive task, a script is better suited:
+
+```python
+# Create a N x 5 grid of connected constant combinators, with every item and their stack size
+
+from draftsman.blueprintable import Blueprint
+from draftsman.constants import Direction
+from draftsman.data import items
+from draftsman.entity import ConstantCombinator
+
+COMBINATOR_HEIGHT = 5
+
+def main():
+    blueprint = Blueprint()
+
+    signal_count = 0
+    signal_index = 0
+    combinators_added = 0
+    x = 0
+    y = 0
+    combinator = ConstantCombinator(direction=Direction.SOUTH)
+
+    # Iterate over every item in order:
+    for item in items.raw:
+        # Ignore hidden items/entities
+        if "flags" in items.raw[item]:
+            if "hidden" in items.raw[item]["flags"]:
+                continue
+        # Keep track of how many signals we've gone through
+        signal_count += 1
+        # Write the stack size signal
+        stack_size = items.raw[item]["stack_size"]
+        combinator.set_signal(signal_index, item, stack_size)
+        signal_index += 1
+        # Once we exceed the number of signals a combinator can hold, place it and reset
+        if signal_index == combinator.item_slot_count:
+            # Add the combinator to the blueprint
+            combinator.id = "{}_{}".format(x, y)
+            blueprint.entities.append(combinator)
+            # Reset the combinator
+            combinators_added += 1
+            y = combinators_added % COMBINATOR_HEIGHT
+            x = int(combinators_added / COMBINATOR_HEIGHT)
+            combinator.set_signals(None)  # Clear signals
+            combinator.tile_position = (x, y)
+            signal_index = 0
+
+    # Add the last combinator if partially full
```

Comment:
Why is there no partially full check needed?
Is it guaranteed to be partially full?

Reply:
Hmm, I suppose you're right actually. This is a bug, though a rare one; the final combinator will always be partially full unless the total number of item signals is perfectly divisible by 20 (or whatever the number of signal slots in the constant combinator is).
This could be fixed by checking the length of the underlying key array:
```python
# This is slightly obscure due to the way the underlying data is structured, but might still be legible enough
# You could also argue the opposite, that this is more illustrative of exactly how Draftsman emulates Factorio
if len(combinator.control_behavior["filters"]) > 0:
    combinator.id = "{}_{}".format(x, y)
    blueprint.entities.append(combinator)
```
Or, because the data is inputted front to back, we can check the first entry to see if it exists: if it does, then we know it has at least one entry that needs to be added:
```python
# This may also seem a little strange if one forgets that the data is added sequentially
if combinator.get_signal(0):
    combinator.id = "{}_{}".format(x, y)
    blueprint.entities.append(combinator)
```
Which one do you think is better?

Reply:
To be honest, the only code I *really* understood was that last snippet, so if you could update Draftsman so that snippet is valid that'd be awesome :)

Reply:
Will, do I'll push it as a patch to version `0.9.2` in a bit

Reply:
Alright, Draftsman is patched, code side works. Keeping this unresolved until the article matches the changes.